Orange/widgets/visualize/utils/widget.py

```diff
@@ -0,0 +1,623 @@
+                    for var in _vars[:max_shown + 2]
+                    if _vars == domain.class_vars
+                    or var not in skip_attrs][:max_shown]
+            if not cols:
+                return ""
+            n_vars = len(_vars)
+            if n_vars > max_shown:
+                cols[-1] = "... and {} others".format(n_vars - max_shown + 1)
+            return \
+                "<b>{}</b>:<br/>".format(singular if n_vars < 2 else plural) \
+                + "<br/>".join(cols)
+
+        domain = self.data.domain
+        parts = (("Class", "Classes", 4, domain.class_vars),
+                 ("Meta", "Metas", 4, domain.metas),
+                 ("Feature", "Features", 10, domain.attributes))
+
+        point_data = self.data[point_id]
+        return "<br/>".join(show_part(point_data, *columns)
+                            for columns in parts)
+
+    def get_tooltip(self, point_ids):
+        """
+        Return the tooltip string for the given points
+
+        The method is called by the plot on mouse hover
+
+        Args:
+            point_ids (list): indices into `data`
+
+        Returns:
+            (str):
+        """
+        text = "<hr/>".join(self._point_tooltip(point_id)
+                            for point_id in point_ids[:MAX_POINTS_IN_TOOLTIP])
+        if len(point_ids) > MAX_POINTS_IN_TOOLTIP:
+            text = "{} instances<hr/>{}<hr/>...".format(len(point_ids), text)
+        return text
+
+    def keyPressEvent(self, event):
+        """Update the tip about using the modifier keys when selecting"""
+        super().keyPressEvent(event)
+        self.graph.update_tooltip(event.modifiers())
+
+    def keyReleaseEvent(self, event):
+        """Update the tip about using the modifier keys when selecting"""
+        super().keyReleaseEvent(event)
+        self.graph.update_tooltip(event.modifiers())
+
+    # Legend
+    def combined_legend(self):
+        """Tells whether the shape and color legends are combined into one"""
+        return self.attr_shape == self.attr_color
+
+
+class OWProjectionWidget(OWProjectionWidgetBase):
+    """Base class for projection widgets."""
+    class Inputs:
+        data = Input("Data", Table, default=True)
+        data_subset = Input("Data Subset", Table)
+
+    class Outputs:
+        selected_data = Output("Selected Data", Table, default=True)
+        annotated_data = Output(ANNOTATED_DATA_SIGNAL_NAME, Table)
+
+    class Warning(OWProjectionWidgetBase.Warning):
+        missing_coords = Msg(
+            "Plot cannot be displayed because data"
+            "is missing for all points")
+
+    class Error(OWProjectionWidgetBase.Error):
+        sparse_data = Msg("Sparse data is not supported")
+
+    settingsHandler = DomainContextHandler()
+    selection = Setting(None, schema_only=True)
+    auto_commit = Setting(True)
+
+    GRAPH_CLASS = OWScatterPlotBase
+    graph = SettingProvider(OWScatterPlotBase)
```

Comment:
I'm not sure how the widget's meta classes work, and nobody's sure how they will work in the future. Adding a setting provider like this can have (or "will can have") side-effects due to processing done to the meta class, which maybe cannot be cancelled in derived classes.

I think it would be better if the base widget wouldn't declare a setting provider and leave this to derived classes.

Reply:
Declaring something as a setting provider should be harmless.

The only thing it does is tell the SettingHandler that there is a "component" named graph that also holds some settings:
https://github.com/biolab/orange3/blob/master/Orange/widgets/settings.py#L125
(called from SettingHandler.create)

Reply:
I would still prefer removing this because the provider would always be replaced. Probably no projection widget will just reuse the base graph --- scatter plot adds a regression line, radviz adds a circle, freeviz adds axes... --- so this provider is always a wrong one.